EventDB: Allocation table

ssardana96 opened the pull request:
Tasks to complete before merging PR:
- [x] IsNft flag added to allocation data. created issue #878 
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master BEFORE merging this PR  
  
Associated PRs (Link as appropriate):
...

ssardana08 committed:
- allocation table gorm struct initialized

ssardana08 committed:
- event emitting - addorOverwriteAllocation

ssardana08 committed:
- error handling in exists

ssardana08 committed:
- emitting add or overwrite events

ssardana08 committed:
- allocation table to storage allocation transformer
- allocation stats handler updated with events db

ssardana08 committed:
- transformation from allocation table to storageNode corrected

ssardana08 committed:
- integration test added (partially)

ssardana08 committed:
- integration test added and tested

ssardana96 commented:
```
ok      0chain.net/smartcontract/dbs/event      0.152s  coverage: 37.2% of statements
```

ssardana96 commented:
Testing on local chain
```
1 | 2022-01-12 07:03:15.230172+00 | 2022-01-12 07:03:15.25912+00 |            | 12ff9665be9e3d18e955d80b77ade9b7d589d8aa23dbfafa1b0714674c5
0da7d | 12ff9665be9e3d18e955d80b77ade9b7d589d8aa23dbfafa1b0714674c50da7d |           1 |             1 | 2147483648 | 1644562994 | [{"blobber
_id":"f65af5d64000c7cd2883f4910eb69086f9d6e6635c744e62afcfab58b938ee25","allocation_id":"12ff9665be9e3d18e955d80b77ade9b7d589d8aa23dbfafa1b07
14674c50da7d","read_price":100000000,"write_price":1000000000,"min_lock_demand":0.1,"max_offer_duration":2678400000000000,"challenge_completi
on_time":120000000000},{"blobber_id":"7a90e6790bcd3d78422d7a230390edc102870fe58c15472073922024985b1c7d","allocation_id":"12ff9665be9e3d18e955
d80b77ade9b7d589d8aa23dbfafa1b0714674c50da7d","read_price":100000000,"write_price":1000000000,"min_lock_demand":0.1,"max_offer_duration":2678
400000000000,"challenge_completion_time":120000000000}] | 9984fb0d9647f57e8aa611a650c4d40fb34a4ad4fca204915fd64a4312fd56d6 | b6c8aea49dddc6f5
0e7e37290f68d8c02c0dc4e4c10370fb963ff0eb1c98e3209370979409e529a9878984383281c032e1839b4e8294ae5ed63e4f48473d5720 | f            |
  0 | 9223372036854775807 |               0 | 9223372036854775807 |                 3600000000000 |              120000000000 | 1641970994 |
f         | f         |         0 |                  0 |          0 |                   0 |          | 2592000000000000 |          0 |
  0 |                0 |               0 |                     0 |                 0 |
  ```

ssardana96 commented:
API change of /allocation
http://localhost:7171/v1/screst/6dba10422e368813802877a85039d3985d96760ed844092319743fb3a76712d7/allocation?allocation=12ff9665be9e3d18e955d80b77ade9b7d589d8aa23dbfafa1b0714674c50da7d

```json
{
"id": "12ff9665be9e3d18e955d80b77ade9b7d589d8aa23dbfafa1b0714674c50da7d",
"tx": "12ff9665be9e3d18e955d80b77ade9b7d589d8aa23dbfafa1b0714674c50da7d",
"data_shards": 1,
"parity_shards": 1,
"size": 2147483648,
"expiration_date": 1644562994,
"blobbers": [
{
"id": "f65af5d64000c7cd2883f4910eb69086f9d6e6635c744e62afcfab58b938ee25",
"url": "http://192.168.29.174:5051",
"geolocation": {
"latitude": 0,
"longitude": 0
},
"terms": {
"read_price": 100000000,
"write_price": 1000000000,
"min_lock_demand": 0.1,
"max_offer_duration": 2678400000000000,
"challenge_completion_time": 120000000000
},
"capacity": 1073741824,
"used": 1073741824,
"last_health_check": 1641970766,
"stake_pool_settings": {
"delegate_wallet": "9984fb0d9647f57e8aa611a650c4d40fb34a4ad4fca204915fd64a4312fd56d6",
"min_stake": 1000000000000,
"max_stake": 1000000000000,
"num_delegates": 50,
"service_charge": 0.3
}
},
{
"id": "7a90e6790bcd3d78422d7a230390edc102870fe58c15472073922024985b1c7d",
"url": "http://192.168.29.174:5052",
"geolocation": {
"latitude": 0,
"longitude": 0
},
"terms": {
"read_price": 100000000,
"write_price": 1000000000,
"min_lock_demand": 0.1,
"max_offer_duration": 2678400000000000,
"challenge_completion_time": 120000000000
},
"capacity": 1073741824,
"used": 1073741824,
"last_health_check": 1641970788,
"stake_pool_settings": {
"delegate_wallet": "9984fb0d9647f57e8aa611a650c4d40fb34a4ad4fca204915fd64a4312fd56d6",
"min_stake": 1000000000000,
"max_stake": 1000000000000,
"num_delegates": 50,
"service_charge": 0.3
}
}
],
"owner_id": "9984fb0d9647f57e8aa611a650c4d40fb34a4ad4fca204915fd64a4312fd56d6",
"owner_public_key": "b6c8aea49dddc6f50e7e37290f68d8c02c0dc4e4c10370fb963ff0eb1c98e3209370979409e529a9878984383281c032e1839b4e8294ae5ed63e4f48473d5720",
"stats": {
"used_size": 0,
"num_of_writes": 0,
"num_of_reads": 0,
"total_challenges": 0,
"num_open_challenges": 0,
"num_success_challenges": 0,
"num_failed_challenges": 0,
"latest_closed_challenge": ""
},
"diverse_blobbers": false,
"preferred_blobbers": null,
"blobber_details": [],
"is_immutable": false,
"read_price_range": {
"min": 0,
"max": 9223372036854776000
},
"write_price_range": {
"min": 0,
"max": 9223372036854776000
},
"max_challenge_completion_time": 3600000000000,
"write_pool_owners": null,
"challenge_completion_time": 120000000000,
"start_time": 1641970994,
"time_unit": 2592000000000000,
"curators": [
""
]
}
```

ssardana08 committed:
- free allocation test corrected

ssardana08 committed:
- /allocations API updated
- old tests fixed

ssardana96 commented:
/allocations API tested
http://localhost:7171/v1/screst/6dba10422e368813802877a85039d3985d96760ed844092319743fb3a76712d7/allocations?client_id=9984fb0d9647f57e8aa611a650c4d40fb34a4ad4fca204915fd64a4312fd56d6
```json
[
{
"id": "765124b91f487320e2e21f8a8c8b788bf9c3f50cbbbcc0bc222e2b83e0717271",
"tx": "765124b91f487320e2e21f8a8c8b788bf9c3f50cbbbcc0bc222e2b83e0717271",
"data_shards": 1,
"parity_shards": 1,
"size": 2147483648,
"expiration_date": 1644587301,
"blobbers": [
{
"id": "7a90e6790bcd3d78422d7a230390edc102870fe58c15472073922024985b1c7d",
"url": "http://192.168.29.174:5052",
"geolocation": {
"latitude": 0,
"longitude": 0
},
"terms": {
"read_price": 100000000,
"write_price": 1000000000,
"min_lock_demand": 0.1,
"max_offer_duration": 2678400000000000,
"challenge_completion_time": 120000000000
},
"capacity": 1073741824,
"used": 1073741824,
"last_health_check": 1641995137,
"stake_pool_settings": {
"delegate_wallet": "9984fb0d9647f57e8aa611a650c4d40fb34a4ad4fca204915fd64a4312fd56d6",
"min_stake": 1000000000000,
"max_stake": 1000000000000,
"num_delegates": 50,
"service_charge": 0.3
}
},
{
"id": "f65af5d64000c7cd2883f4910eb69086f9d6e6635c744e62afcfab58b938ee25",
"url": "http://192.168.29.174:5051",
"geolocation": {
"latitude": 0,
"longitude": 0
},
"terms": {
"read_price": 100000000,
"write_price": 1000000000,
"min_lock_demand": 0.1,
"max_offer_duration": 2678400000000000,
"challenge_completion_time": 120000000000
},
"capacity": 1073741824,
"used": 1073741824,
"last_health_check": 1641995113,
"stake_pool_settings": {
"delegate_wallet": "9984fb0d9647f57e8aa611a650c4d40fb34a4ad4fca204915fd64a4312fd56d6",
"min_stake": 1000000000000,
"max_stake": 1000000000000,
"num_delegates": 50,
"service_charge": 0.3
}
}
],
"owner_id": "9984fb0d9647f57e8aa611a650c4d40fb34a4ad4fca204915fd64a4312fd56d6",
"owner_public_key": "b6c8aea49dddc6f50e7e37290f68d8c02c0dc4e4c10370fb963ff0eb1c98e3209370979409e529a9878984383281c032e1839b4e8294ae5ed63e4f48473d5720",
"stats": {
"used_size": 0,
"num_of_writes": 0,
"num_of_reads": 0,
"total_challenges": 0,
"num_open_challenges": 0,
"num_success_challenges": 0,
"num_failed_challenges": 0,
"latest_closed_challenge": ""
},
"diverse_blobbers": false,
"preferred_blobbers": null,
"blobber_details": [
{
"blobber_id": "7a90e6790bcd3d78422d7a230390edc102870fe58c15472073922024985b1c7d",
"allocation_id": "765124b91f487320e2e21f8a8c8b788bf9c3f50cbbbcc0bc222e2b83e0717271",
"size": 0,
"allocation_root": "",
"write_marker": null,
"stats": null,
"terms": {
"read_price": 100000000,
"write_price": 1000000000,
"min_lock_demand": 0.1,
"max_offer_duration": 2678400000000000,
"challenge_completion_time": 120000000000
},
"min_lock_demand": 0,
"spent": 0,
"penalty": 0,
"read_reward": 0,
"returned": 0,
"challenge_reward": 0,
"final_reward": 0,
"challenge_pool_integral_value": 0
},
{
"blobber_id": "f65af5d64000c7cd2883f4910eb69086f9d6e6635c744e62afcfab58b938ee25",
"allocation_id": "765124b91f487320e2e21f8a8c8b788bf9c3f50cbbbcc0bc222e2b83e0717271",
"size": 0,
"allocation_root": "",
"write_marker": null,
"stats": null,
"terms": {
"read_price": 100000000,
"write_price": 1000000000,
"min_lock_demand": 0.1,
"max_offer_duration": 2678400000000000,
"challenge_completion_time": 120000000000
},
"min_lock_demand": 0,
"spent": 0,
"penalty": 0,
"read_reward": 0,
"returned": 0,
"challenge_reward": 0,
"final_reward": 0,
"challenge_pool_integral_value": 0
}
],
"is_immutable": false,
"read_price_range": {
"min": 0,
"max": 9223372036854776000
},
"write_price_range": {
"min": 0,
"max": 9223372036854776000
},
"max_challenge_completion_time": 3600000000000,
"write_pool_owners": null,
"challenge_completion_time": 120000000000,
"start_time": 1641995301,
"time_unit": 2592000000000000,
"curators": [
""
]
}
]
```

ssardana96 commented:
New API
http://localhost:7171/v1/screst/6dba10422e368813802877a85039d3985d96760ed844092319743fb3a76712d7/active_allocations_count
```json
{
"active_allocations_count": 1
}
```

http://localhost:7171/v1/screst/6dba10422e368813802877a85039d3985d96760ed844092319743fb3a76712d7/blobber_allocations_count
```json
{
"blobber_allocations_count": 2
}
```

ssardana08 committed:
- new APIs added

kushthedude commented on code/go/0chain.net/smartcontract/dbs/event/allocation_test.go:
```diff
@@ -0,0 +1,398 @@
+		Signature              string           `json:"signature"`
+	}
+
+	type BlobberAllocation struct {
+		BlobberID       string                  `json:"blobber_id"`
+		AllocationID    string                  `json:"allocation_id"`
+		Size            int64                   `json:"size"`
+		AllocationRoot  string                  `json:"allocation_root"`
+		LastWriteMarker *WriteMarker            `json:"write_marker"`
+		Stats           *StorageAllocationStats `json:"stats"`
+		Terms           Terms                   `json:"terms"`
+		// MinLockDemand for the allocation in tokens.
+		MinLockDemand state.Balance `json:"min_lock_demand"`
+		Spent         state.Balance `json:"spent"`
```
Do we need to define them all again here in tests?

ssardana96 replied:
Yes. else, cyclic dependency

kushthedude replied:
@Sriep @ssardana08 Should we plan to abstract out the models into a different package to avoid re-declaration and remove the redundancy from code? This way we need to maintain two models out of which one is for ensuring if tests are working which is against the DRY.

ssardana96 replied:
IMO we should do this.
Models should be abstracted out in a different package.

Sriep replied:
There is this issue https://github.com/0chain/0chain/issues/522.

kushthedude commented on code/go/0chain.net/smartcontract/storagesc/allocation.go:
```diff
@@ -607,6 +613,11 @@ func (sc *StorageSmartContract) saveUpdatedAllocation(all *StorageNodes,
 
+	err = emitAddOrOverwriteAllocation(alloc, balances)
+	if err != nil {
+		return
```
No error wrapping here?

ssardana08 replied:
will add

ssardana96 replied:
that's happening on it's parent function `updateAllocationRequestInternal`
just error return should be good.

kushthedude commented on code/go/0chain.net/smartcontract/storagesc/allocation_eventdb.go:
```diff
@@ -0,0 +1,239 @@
+		ReadPriceRange:             PriceRange{alloc.ReadPriceMin, alloc.ReadPriceMax},
+		WritePriceRange:            PriceRange{alloc.WritePriceMin, alloc.WritePriceMax},
+		MaxChallengeCompletionTime: time.Duration(alloc.MaxChallengeCompletionTime),
+		// todo: to be added with WritePool : select user_id from WritePools where allocation_id = ?
```
Can we create an issue for this?

ssardana96 replied:
Will create one

kushthedude commented on code/go/0chain.net/smartcontract/storagesc/handler.go:
```diff
@@ -120,7 +120,7 @@ func (msc *StorageSmartContract) GetTransactionByHashHandler(
 }
 
-func (ssc *StorageSmartContract) GetAllocationsHandler(ctx context.Context,
+func (ssc *StorageSmartContract) GetAllocationsHandlerDeprecated(ctx context.Context,
```
Can we add a line above it starting with `Deprecated:`, in this way static linter can give warning with deprecation ?

ssardana08 committed:
- deprecated comment added

Sriep commented on code/go/0chain.net/smartcontract/storagesc/allocation_eventdb.go:
```diff
@@ -0,0 +1,239 @@
+	"time"
+)
+
+func allocationTableToStorageAllocation(alloc *event.Allocation, balances cstate.StateContextI) (*StorageAllocation, error) {
```
Pass an `*event.EventDb` object instead of cstate.StateContextI object.

ssardana96 replied:
done

Sriep commented on code/go/0chain.net/smartcontract/storagesc/allocation_eventdb.go:
```diff
@@ -0,0 +1,239 @@
+	return sa, nil
+}
+
+func getClientAllocationsFromDb(clientID string, balances cstate.StateContextI) ([]*StorageAllocation, error) {
```
Pass an `*event.EventDb` object instead of cstate.StateContextI object.

ssardana96 replied:
done

Sriep reviewed (changes requested):
System tests need to pass. Their failure suggests that data returned from `GetAllocationsHandler` has changed where there should be no change. This should be a block box Pr invisible to current system tests.

Due to backward compatibility issues, merging is blocked by https://github.com/0chain/0chain/issues/857. This will likely result in changes to `storagesc/handler.go`.

When finished we will need to create a [new system test issue](https://github.com/0chain/system_test/issues) to test the two new endpoints `active_allocations_count` and `blobber_allocations_count`.

ssardana08 committed:
- replaced StateContextI with eventdb as not required

service-0chain commented:
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | feat/allocation_table  |
| blobber  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

ssardana08 committed:
- nil eventdb error handling

ssardana08 committed:
- adding emitting of allocation at curatorTransferAllocation to transfer owner

ssardana08 committed:
- mocks updated for TestTransferAllocation

ssardana08 committed:
- handling curators better

ssardana08 committed:
- added allocation update to places missing like curator, challenges, blobber, writemarker

ssardana08 committed:
- curator mocks updated with EmitEvent()

ssardana08 committed:
- curator mocks updated with EmitEvent()

ssardana08 committed:
- emit add/overwrite allocation updated

ssardana08 committed:
- curators key removed from allocation table
- storage allocation curator will be populated with curator table
- changes to allocation removed curator.go

iamrz1 committed:
Fetching curator IDs from curators table using allocation ID

iamrz1 committed:
Added foreign key and constraints on allocation id from allocation table

iamrz1 committed:
Fix test allocation table

ssardana08 committed:
- MaxNum Delegates fixed

iamrz1 committed:
Fixed number of endpoints in handler test

iamrz1 committed:
Fixed upload issue with new allocation

ssardana08 committed:
- build ossue

ssardana96 commented:
completed in #1054
